### DISCUSSION TITLE
Brainstorm: IPVM Demos

### DISCUSSION THREAD

**Discussion by depatchedmode** [ipvm-wg/homestar#295] [Ideas]
Let's collect ideas for quick demos we can throw together that will (in no particular order):
1. Test IPVM's capabilities in practice
2. Demonstrate IPVM's capabilities to other folks — with an emphasis on the novel ones
3. Identify the good, bad & ugly bits of the DX
4. Provide interested teams with a variety of stable starting points to experiment from
5. Solicit feedback on IPVM that will help us plan & prioritize

These demos should be sized to fit in a 2 week sprint. The demos should make use of existing IPVM capabilities, or capabilities that are expected to land in the coming quarter. Seems natural that we'll start with simple demos and start to compose them into more complex demos over time.

**Comment by bgins** [comment]
Here are a couple of ideas.

### Analyze an IPVM workflow with an IPVM workflow

We are adding tracing to IPVM Homestar to collect events that happen when running an IPVM workflow. We can extract this data and derive metrics from it.

One interesting demo would be to run an IPVM workflow that analyzes event data and derives metrics from it. This demo would be interesting because it shows an example use case _and_ it teaches about IPVM workflows at the same time.

### Network analyzer and visualizer

We plan to collect data on the nodes participating an IPVM workflow. It might be interesting to build a network analyzer and visualizer that displays information on the nodes while a workflow is in progress.

For each node, we might display:

- CPU
- Memory
- Currently running which task
- Blocked, awaiting some other task to complete
- Tasks executed

Likely more. There may also be some meta information, such as number or nodes and latency between nodes that we could display.

**Reply by expede** [reply]
Recursive IPVM 😁 I love everything about this

**Comment by bgins** [comment]
## Ray tracing

Ray tracing is a technique used in rendering digital images. Parallel ray tracing renders in parallel on one machine or across multiple machines. As a reference point, Tachyon is one popular implementation: https://en.wikipedia.org/wiki/Tachyon_(software).

The more parallelism and time spent on a ray trace, the better quality the render. We could run workflows where we render across multiple machines with varied time constraints to show how can renders can be improved through collaboration.

This idea is a neat technical demo, but also has practical applications. For example, the auto industry uses ray tracing to render car concepts. Likely there other other industries that also use this technique.
